### DISCUSSION TITLE
not possible to nonblock read from stdin as a windows executable launched via ssh

### DISCUSSION THREAD

**Discussion by delyas** [microsoft/WSL#6807] [Q&A]
### Windows Build Number

Microsoft Windows [Version 10.0.18363.1443]

### WSL Version

WSL 1

### Distro Version

Ubuntu 20.04

### Other Software

openssh-server
my custom application reading from stdin
(https://github.com/GENIVI/ramses/blob/26311b24133e3f927f650542af05da0c53ca5e46/framework/PlatformAbstraction/src/ConsoleInput.cpp#L59)

### Repro Steps

setup wsl to be able to execute windows executables via ssh
launch mentioned application via ssh and write something to stdin

### Expected Behavior

If I run the executable running the linked code via ssh, stdin will be of type FILE_TYPE_PIPE. I can PeekNamedPipe on the stdin file to see if there is something to read, or in an alternative implementation I can wait for it to go to a signaled state with WaitForSingleObject et al and then read from it.

### Actual Behavior

If I run the quoted code via ssh, stdin will be of type FILE_TYPE_PIPE. There is no way for me to read this pipe in a non blocking way:

- In the linked implementation PeekNamedPipe on the stdin handle will fail with ERROR_INVALID_FUNCTION.
- Just calling ReadFile on it works, but blocks - blocks not only this thread (which would be okayish), but also another thread trying to open a window to render to, so it is not an acceptable solution.
- Calling ReadFileEx followed by WaitForSingleObjectEx or MsgWaitForMultipleObjects behaves the same, because also ReadFileEx blocks, most likely because stdin has no FILE_FLAG_OVERLAPPED flag.
- Doing an initial WaitForSingleObject on the stdin handle will return WAIT_OBJECT_0 immediately without anything to read. MsgWaitForMultipleObjects behaves the same, even if I specify 0 as dwWakeMask.

I love WSL, I really do. The fact that I have come this far remotely launching a graphical application the same way I would on a linux machine is just amazing. It would be a shame if it would fail here, at this very last step. 
Please help me to find a path to victory here.

**Comment by delyas** [comment]
Answering to my own discussion (which wasn't a discussion, actually): If this is not possible currently, could we make a feature request out of this, @therealkenc ?

**Comment by therealkenc** [comment]
There was not enough here to determine what you are trying to do and form an actionable.

**Comment by delyas** [comment]
Changing one of the following three (mis)behaviors would allow me to non blocking read from a FILE_TYPE_PIPE stdin:

1. 
Expected: I can wait with MsgWaitForMultipleObjects or WaitForSingleObject for something to read from that pipe, before calling ReadFile.
Actual: Pipe is always in a signalled state, even if empty, which makes me call a blocking ReadFile.

2.
Expected: Overlapped file operations allow to work asynchronously with this pipe. 
Actual: ReadFileEx blocks until something can be read.

3. 
Expected: PeekNamedPipe will work on this pipe. I can call it to test if there is something to read before calling ReadFile.
Actual: PeekNamedPipe fails with ERROR_INVALID_FUNCTION

**Reply by therealkenc** [reply]
If you like, open new submission with CLI repro steps that evoke the `ERROR_INVALID_FUNCTION`.
